File: java/tutorial/ch/softappeal/yass/tutorial/py3/SocketServer.java
package ch.softappeal.yass.tutorial.py3;

import ch.softappeal.yass.core.Interceptor;
import ch.softappeal.yass.transport.MessageSerializer;
import ch.softappeal.yass.transport.socket.SimpleSocketBinder;
import ch.softappeal.yass.transport.socket.SimpleSocketTransport;
import ch.softappeal.yass.tutorial.acceptor.socket.SocketAcceptor;
import ch.softappeal.yass.tutorial.contract.EchoService;
import ch.softappeal.yass.tutorial.contract.Instrument;
import ch.softappeal.yass.tutorial.contract.Logger;
import ch.softappeal.yass.tutorial.contract.SystemException;
import ch.softappeal.yass.tutorial.contract.instrument.InstrumentService;
import ch.softappeal.yass.util.Exceptions;
import ch.softappeal.yass.util.NamedThreadFactory;
import ch.softappeal.yass.util.Nullable;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.Executor;
import java.util.concurrent.Executors;

import static ch.softappeal.yass.tutorial.contract.Config.ACCEPTOR;

public final class SocketServer {

    private static final class EchoServiceImpl implements EchoService {
        @Override public @Nullable Object echo(final @Nullable Object value) {
            if ("exception".equals(value)) {
                throw new SystemException("exception");
            }
            return value;
        }
    }

    private static final class InstrumentServiceImpl implements InstrumentService {
        @Override public List<Instrument> getInstruments() {
            return new ArrayList<>();
        }
        @Override public void showOneWay(final boolean testBoolean, final int testInt) {
            // empty
        }
    }

    public static void main(final String... args) {
        final Executor executor = Executors.newCachedThreadPool(new NamedThreadFactory("executor", Exceptions.STD_ERR));
        final Interceptor logger = new Logger(null, Logger.Side.SERVER);
        new SimpleSocketTransport(
            executor,
            new MessageSerializer(SocketClient.SERIALIZER),
            new ch.softappeal.yass.core.remote.Server(
                ACCEPTOR.echoService.service(new EchoServiceImpl()),
                ACCEPTOR.instrumentService.service(new InstrumentServiceImpl(), logger)
            )
        ).start(executor, new SimpleSocketBinder(SocketAcceptor.ADDRESS));
        System.out.println("started");
    }

}

File: py3/tutorial/socket_server.py
import socket
from ssl import SSLContext, SSLSocket

from typing import Any, List, cast

from tutorial.base_types_external import Integer
from tutorial.generated import contract, ACCEPTOR
from tutorial.generated.contract import EchoService, SystemException
from tutorial.generated.contract.instrument import InstrumentService
from tutorial.socket_client import serializer, address, serverPrinter, SocketStream, sslContext
from yass import Server, defaultServerTransport, ServerTransport


class EchoServiceImpl(EchoService):
    def echo(self, value: Any) -> Any:
        if value == "exception":
            e = SystemException()
            e.details = "exception"
            raise e
        return value


class InstrumentServiceImpl(InstrumentService):
    def getInstruments(self) -> List[contract.Instrument]:
        return []

    def showOneWay(self, testBoolean: bool, testInt: Integer) -> None:
        pass


def socketServer(transport: ServerTransport, address: Any, backlog: int, sslContext: SSLContext) -> None:
    serverSocket = socket.socket()
    serverSocket.bind(address)
    serverSocket.listen(backlog)
    while True:
        try:
            s = cast(SSLSocket, sslContext.wrap_socket(serverSocket.accept()[0], server_side=True))
            try:
                print(s.getpeercert())
                transport.invoke(SocketStream(s))
            finally:
                s.close()
        except Exception as e:
            print("exception:", e)


if __name__ == "__main__":
    print("started")
    socketServer(
        defaultServerTransport(
            serializer,
            Server([
                ACCEPTOR.echoService.service(EchoServiceImpl()),
                ACCEPTOR.instrumentService.service(InstrumentServiceImpl(), serverPrinter)
            ])
        ),
        address,
        5,
        sslContext("ClientCA.cert.pem", "Server.key.pem")
    )
